File: src/race_overlay/alignment.py
from datetime import timedelta

from race_overlay.models import ActivityTrack, ClipAlignment, VideoClip


def align_clip(
    activity: ActivityTrack,
    clip: VideoClip,
    global_offset_seconds: float,
    per_video_offset_seconds: float,
) -> ClipAlignment:
    clip_start = clip.creation_time + timedelta(seconds=global_offset_seconds + per_video_offset_seconds)
    clip_end = clip_start + timedelta(seconds=clip.duration_seconds)
    activity_start = activity.samples[0].timestamp
    activity_end = activity.samples[-1].timestamp
    overlay_start = max(clip_start, activity_start)
    overlay_end = min(clip_end, activity_end)
    if clip_end <= activity_start or clip_start >= activity_end:
        return ClipAlignment(clip, "outside", clip_start, clip_end, None, None)
    if clip_start < activity_start or clip_end > activity_end:
        return ClipAlignment(clip, "partial", clip_start, clip_end, overlay_start, overlay_end)
    return ClipAlignment(clip, "inside", clip_start, clip_end, overlay_start, overlay_end)

File: web/src/runtime/alignment.ts
import type { ActivityTrack, ClipAlignment, VideoClip } from "./models";
import { addSeconds, maxIso, minIso } from "./time";

export interface AlignClipOptions {
  globalOffsetSeconds: number;
  perVideoOffsetSeconds: number;
}

export function alignClip(activity: ActivityTrack, clip: VideoClip, options: AlignClipOptions): ClipAlignment {
  if (activity.samples.length < 2) {
    throw new Error("activity must contain at least 2 samples");
  }
  const clipStart = addSeconds(clip.creationTime, options.globalOffsetSeconds + options.perVideoOffsetSeconds);
  const clipEnd = addSeconds(clipStart, clip.durationSeconds);
  const activityStart = activity.samples[0].timestamp;
  const activityEnd = activity.samples[activity.samples.length - 1].timestamp;
  const overlayStart = maxIso(clipStart, activityStart);
  const overlayEnd = minIso(clipEnd, activityEnd);

  if (new Date(clipEnd).getTime() <= new Date(activityStart).getTime() || new Date(clipStart).getTime() >= new Date(activityEnd).getTime()) {
    return { clip, status: "outside", clipStart, clipEnd, overlayStart: null, overlayEnd: null };
  }
  if (new Date(clipStart).getTime() < new Date(activityStart).getTime() || new Date(clipEnd).getTime() > new Date(activityEnd).getTime()) {
    return { clip, status: "partial", clipStart, clipEnd, overlayStart, overlayEnd };
  }
  return { clip, status: "inside", clipStart, clipEnd, overlayStart, overlayEnd };
}
